benchmarks/libseff/generator/main.c

```diff
@@ -0,0 +1,142 @@
+#include "seff.h"
+#include <stdio.h>
+
+DEFINE_EFFECT(yield, 0, void, { int64_t value; });
+
+// Tree data structure
+struct node_t {
+  struct node_t* l;
+  int64_t v;
+  struct node_t* r;
+};
+typedef struct node_t node_t;
+typedef struct node_t* tree_t;
+
+// Generator data structure
+struct thunk_t {
+  int64_t v;
+  struct thunk_t* next;
+};
+typedef struct thunk_t thunk_t;
+typedef thunk_t* generator_t;
+
+// Allocate and free a tree node
+static inline node_t* allocNode() { return malloc(sizeof(node_t));}
+static inline void freeNode(node_t* node) { free(node); }
+
+// Allocate a tree on the heap
+static tree_t makeTree(int n) {
+  if (n == 0) return NULL;
+  tree_t node = allocNode();
+  tree_t t = makeTree (n - 1);
+
+  // Note: both left and right pointers point to the same memory, 
+  // to match the other languages
+  node->l = t;
+  node->r = t;
+  node->v = n;
+  return node;
+}
```

Comment:
I feel we're cheating here a bit, the reason why other languages can point to the same subtree is because the combination of immutability + GC allows them to perform shallow copies that won't leak.
In C, we sadly can't do that easily (maybe ref counting, but out of scope), which makes the freeing of this tree very specialized (`freeTree` only frees trees created by `makeTree`, not general binary trees).

I'll summon the chairs here (@phischu @jasigal ), but I feel we have to either bite the bullet, allocate each subtree independently and have the generator benchmark be slower than others for reasons unrelated to EH, or reconsider this benchmark altogether, maybe having a tree that is actually general with different values on each node, forcing implementations to actually create the whole tree in memory.

Reply:
Good points, it's hard to say in my opinion what the best option is. I'm tempted to say that the C program shouldn't free the tree because the program ends immediately, and thus the Haskell and OCaml code probably don't explicity free the space anyway.

I think the change which would make the work consistent across all languages is your suggestion of having distinct subtrees. Then we would be benchmarking the EH factor of the code.

Reply:
> Good points, it's hard to say in my opinion what the best option is. I'm tempted to say that the C program shouldn't free the tree because the program ends immediately, and thus the Haskell and OCaml code probably don't explicity free the space anyway.

I also feel like cheating there a bit, but it's a matter of whether we write benchmarks to be similar (using OCaml as a reference) or as good as possible for language X.
 
> I think the change which would make the work consistent across all languages is your suggestion of having distinct subtrees. Then we would be benchmarking the EH factor of the code.

Yes, however, it'll probably benchmark how quickly you can construct a tree rather than traverse it (but I'm speculating)

Reply:
> I'm tempted to say that the C program shouldn't free the tree because the program ends immediately, and thus the Haskell and OCaml code probably don't explicity free the space anyway.

I agree. I consider both of the following options to be fair:
- What you do right now, i.e. `freeTree` that works on this special kind of tree
- Don't free it at all, because the program ends anyway

Reply:
Okay, I will change it so that it doesn't free the memory after all, since it most likely mirrors what the other languages do. Even though it leaks, the program exits anyway.

Reply:
I am advocating for the opposite: I would consider a memory-leaking (or any resource-leaking) benchmark as an invalid benchmark. Continuation cancellation is a key facility in real world programming with effect handlers.

Reply:
That makes sense! Okay, I will revert to having `freeTree` :)